continuationsPlugin/src/main/scala/continuations/DefDefTransforms.scala

```diff
@@ -0,0 +1,65 @@
+package continuations
+
+import dotty.tools.dotc.ast.Trees.*
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.ast.tpd.TypedTreeCopier
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Flags
+import dotty.tools.dotc.core.Names
+import dotty.tools.dotc.core.Symbols
+import dotty.tools.dotc.core.Types
+import dotty.tools.dotc.core.Types.Type
+
+object DefDefTransforms {
+
+  def transformNonSuspending(tree: tpd.DefDef, cpy: TypedTreeCopier)(
+      using ctx: Context): tpd.DefDef =
+    tree match {
+      case DefDef(name, paramss, tpt, rhs) =>
+        val suspendClazz = Symbols.requiredClass("continuations.Suspend")
+        val completion: tpd.ValDef = tpd.ValDef(
+          Symbols
+            .newSymbol(
+              tree.symbol,
+              Names.termName("completion"),
+              Flags.LocalParam,
+
+              Symbols.requiredPackage("continuations").requiredType("Continuation").typeRef.appliedTo(Types.OrType(tpt.tpe, ctx.definitions.AnyType, false))
+            )
+            .asTerm,
+          rhs = theEmptyTree
+        )
+        val params: List[tpd.ParamClause] = if (paramss.isEmpty) {
+          List(List(completion))
+        } else
+          tree.paramss.zipWithIndex.map { (pc, i) =>
+            val newPc: tpd.ParamClause = pc
+              .filterNot { param =>
+                param match {
+                  case p: ValDef[Type] =>
+                    p.typeOpt.hasClassSymbol(suspendClazz) && p
+                      .symbol
+                      .flags
+                      .isOneOf(Flags.GivenOrImplicit)
+                  case t: TypeDef[Type] =>
+                    t.typeOpt.hasClassSymbol(suspendClazz) && t
+                      .symbol
+                      .flags
+                      .isOneOf(Flags.GivenOrImplicit)
+                }
+              }
+              .asInstanceOf[tpd.ParamClause]
+            if (i == 0) {
+              newPc.appended(completion).asInstanceOf[tpd.ParamClause]
+            } else newPc
+          }.filterNot(_.isEmpty)
+        cpy.DefDef(tree)(paramss = params, tpt = tpd.TypeTree(ctx.definitions.ObjectClass.typeRef))
+    }
+
+  def transformSuspendedDef(tree: tpd.DefDef, cpy: TypedTreeCopier)(
+      using ctx: Context): tpd.DefDef = tree match {
+    case BodyHasNoSuspensionPoint(_) => cpy.DefDef(tree)()
```

Comment:
Do we still generate the Continuation arg if a suspended function has no suspension points?

Reply:
The kotlin `coroutines-test` does, and I think we need to if we expect to call `run` at the end of the world.